### PR TITLE
ENH: improve MinariDataset

### PR DESCRIPTION
**Pull request by younik**
# Description

Add `__getitem__`and `__len__()` to `MinariDataset` and few refactors.

## Type of change

- New feature (non-breaking change which adds functionality)

# Checklist:

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files` (see `CONTRIBUTING.md` instructions to set it up)
- [x] I have run `pytest -v` and no errors are present.
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I solved any possible warnings that `pytest -v` has generated that are related to my code to the best of my knowledge.
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: docs/tutorials/q_learning_torch.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+# 
+# Where `T` is the toal number of step in the episode. However, we have a batch of episodes with different length `T`; we use ``mask`` to zeros the loss after the episode is terminated. 
+# 
+# Lastly, we can assess the performance of the agent that utilizes the learned `Q`-values:
+
+env = gym.make("CartPole-v1", render_mode="human")
+obs, _ = env.reset(seed=42)
+done = False
+accumulated_rew = 0
+while not done:
```
[the remark this answers is not in the document]
Comment (balisujohn) [reply docs/tutorials/q_learning_torch.py:169]: For reference, I think the code is not fully deterministic since I got:
````
Offline Q-Learning accumulated rew:  160.0
Dataset mean accumulated rew:  22.12796
````

**Review by balisujohn — COMMENTED** [review]
Seems to run fine and produces a performant policy, I requested a few minor changes, and asked a few questions.


One more comment: It might be good to provide a requirements.txt for tutorials(though the version of Minari this will depend on isn't in a release yet, so I suppose it would have to be in a later PR)

File: minari/dataset/minari_dataset.py [hunk 2]
```diff
@@ -386,3 +381,11 @@ def update_dataset_from_buffer(self, buffer: List[dict]):
 
     def __iter__(self):
         return self.iterate_episodes()
+
+    def __getitem__(self, idx: int) -> EpisodeData:
```
Comment (balisujohn) [inline minari/dataset/minari_dataset.py:385]: All changes look good, but please add  tests for `__iter__ `,`__getitem__` and `__len__`

File: docs/tutorials/q_learning_torch.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+# %%%
+# We use the Mean Square Error loss, as a classic regression task. In fact, we want to predict the `Q`-values of each state-action pairs that we take from the trajectories. 
+# The training loop will look like this:
+
+num_epochs = 3
+gamma = 0.99
+
+for epoch in range(num_epochs):
+    for batch in tqdm(dataloader):
+        q_values = q_net(batch['observations'])
```
Comment (Howuhh) [inline docs/tutorials/q_learning_torch.py:134]: I think it would be good to add shape annotations here, so that users can understand loss completely

File: docs/tutorials/q_learning_torch.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        q_targets = batch['rewards'].float() + gamma * torch.max(q_values[:, 1:], dim=-1).values
+        q_pred = q_values[:, :-1].gather(-1, batch['actions'].unsqueeze(-1)).squeeze()
+            
+        mask = 1 - torch.cumsum(
+            torch.logical_or(batch["terminations"], batch["truncations"]),
+            dim=-1
+        )
+        loss = loss_fn(q_pred * mask, q_targets * mask)
```
[the remark this answers is not in the document]
Comment (younik) [reply docs/tutorials/q_learning_torch.py:142]: I think you are right as well; and yes, if I change it, it performs worse

File: docs/tutorials/q_learning_torch.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+        loss.backward()
+        optimizer.step()
+
+    print(f"Epoch: {epoch}/{num_epochs}, Loss: {loss.item()}")
+
+# %%%
+# where we compute the ``q_targets`` using the SARSA objective 
+# 
+# .. math ::
+#   r(s_t, a_t) + \gamma \max_{a' \in \mathcal{A}} \hat{Q}_{\theta}(s_{t+1}, a')
```
[the remark this answers is not in the document]
Comment (younik) [reply docs/tutorials/q_learning_torch.py:154]: The initial idea of this tutorial was to show how to integrate the dataset in a common PyTorch pipeline bus, afaik, there is no paper showing this very simple method (that doesn't work in general). I wanted this to be as simple as possible

Maybe I should update the tutorial to reimplement a paper; it will be more complicated, but also more informative

File: docs/tutorials/q_learning_torch.py (hunk 3 above)
Comment (younik) [reply docs/tutorials/q_learning_torch.py:134]: Good point, I will update

File: minari/dataset/minari_dataset.py (hunk 2 above)
Comment (younik) [reply minari/dataset/minari_dataset.py:385]: Much easier to add after we merge https://github.com/Farama-Foundation/Minari/pull/101

File: docs/tutorials/q_learning_torch.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (balisujohn) [reply docs/tutorials/q_learning_torch.py:154]: If this method is custom and there is no paper describing it, I think that's fine.

File: docs/tutorials/behavioral_cloning.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+"""
+# %%%
+# We present here how to perform behavioral cloning on a Minari dataset using PyTorch.
+# We will start generating the dataset of the expert policy for the `CartPole-v1 <https://gymnasium.farama.org/environments/classic_control/cart_pole/>`_ environment, which is a classic control problem.
+
+# %%
+# Policy training
+# ~~~~~~~~~~~~~~~~~~~
+# To train the expert policy, we use the library rl_zoo3.
+# After installing the lirbary with ``pip install rl_zoo3``,
```
Comment (rodrigodelazcano) [inline docs/tutorials/behavioral_cloning.py:14]: "library"

File: docs/index.md [hunk 7]
```diff
@@ -25,6 +25,7 @@ api/data_collector
 :glob:
 :caption: Tutorials
 tutorials/**/index
+tutorials/behavioral_cloning
```
Comment (rodrigodelazcano) [inline docs/index.md:28]: I think we can include this tutorial in a new tab called "Using Datasets". What do you think?

**Review by rodrigodelazcano — CHANGES_REQUESTED** [review]
I just added some suggestions

File: docs/tutorials/behavioral_cloning.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+#
+# ``python -m rl_zoo3.train --algo ppo --env CartPole-v1``
+
+# %%
+# This will generate a new folder named `log` with the expert policy.
+
+# %%
+# Imports
+# ~~~~~~~~~~~~~~~~~~~
+# Let's import all the occurrences and set the random seed for reproducibility:
```
Comment (balisujohn) [inline docs/tutorials/behavioral_cloning.py:25]: maybe change "occurences" to "required packages"

File: docs/tutorials/behavioral_cloning.py [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+    action = policy_net(torch.Tensor(obs)).argmax()
+    obs, rew, ter, tru, _ = env.step(action.numpy())
+    done = ter or tru
+    accumulated_rew += rew
+
+env.close()
+print("Accumulated rew: ", accumulated_rew)
+
+# %%
+# We can visually observe that the learned policy aces this simple control task, and we get the maximum reward.
```
Comment (balisujohn) [inline docs/tutorials/behavioral_cloning.py:184]: this should say what a good reward is for the environment, I got 500 as the accumulated reward when I ran this(Which I assume is a good reward).

**Review by balisujohn — COMMENTED** [review]
Really minor change requests, and a reminder that we should add a version locked requirements.txt after the next minari release. Otherwise, looks good, and ready to merge, assuming 500 is a good amount of total reward for the final cloned policy.

File: docs/tutorials/behavioral_cloning.py (hunk 9 above)
Comment (younik) [reply docs/tutorials/behavioral_cloning.py:184]: 500 is the maximum obtainable, since after the episode is truncated. I add the explanation

File: docs/index.md (hunk 7 above)
Comment (younik) [reply docs/index.md:28]: Sounds good, done

**Review by balisujohn — APPROVED** [review]
LGTM :^)

**Review by rodrigodelazcano — APPROVED** [review]

**Comment by rodrigodelazcano** [comment]
Looks good! Ready to merge
